File: api/csv/github.js
/**
 * Vercel serverless function that:
 *  - Creates a new branch off base (default: main)
 *  - Commits the posted birthdays.csv to that branch
 *  - Opens a Pull Request back to base
 *
 * Env vars (set in Vercel Project Settings -> Environment Variables):
 *   REQUIRED:
 *     - GITHUB_APP_ID
 *     - GITHUB_APP_INSTALLATION_ID
 *     - GITHUB_APP_PRIVATE_KEY  (full PEM string inc. BEGIN/END, with newlines)
 *   OPTIONAL (defaults shown):
 *     - GITHUB_REPO_OWNER=grinwi
 *     - GITHUB_REPO=birth-app
 *     - GITHUB_BRANCH=main
 *     - GITHUB_FILE_PATH=birthdays.csv
 *     - PR_BRANCH_PREFIX=update-birthdays
 *
 * Request:
 *   POST text/csv body (raw CSV)
 *
 * Response:
 *   200 JSON { pr_url, pr_number, branch } on success
 */
const { App } = require('@octokit/app');

function setCors(res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
}

module.exports = async (req, res) => {
  setCors(res);

  if (req.method === 'OPTIONS') {
    return res.status(204).end();
  }
  if (req.method !== 'POST') {
    return res.status(405).send('Method Not Allowed');
  }

  try {
    const csv = await readRawBody(req);
    if (!csv || typeof csv !== 'string' || csv.length < 10) {
      return res.status(400).send('Invalid or empty CSV');
    }

    const owner = process.env.GITHUB_REPO_OWNER || 'grinwi';
    const repo = process.env.GITHUB_REPO || 'birth-app';
    const baseBranch = process.env.GITHUB_BRANCH || 'main';
    const filePath = process.env.GITHUB_FILE_PATH || 'birthdays.csv';
    const prBranchPrefix = (process.env.PR_BRANCH_PREFIX || 'update-birthdays').replace(/[^a-zA-Z0-9._/-]/g, '-') || 'update-birthdays';

    const appId = process.env.GITHUB_APP_ID;
    const installationId = process.env.GITHUB_APP_INSTALLATION_ID;
    const privateKey = process.env.GITHUB_APP_PRIVATE_KEY;

    if (!appId || !installationId || !privateKey) {
      return res.status(500).send('Server missing GITHUB_APP_ID / GITHUB_APP_INSTALLATION_ID / GITHUB_APP_PRIVATE_KEY');
    }

    const app = new App({ appId, privateKey });
    const octokit = await app.getInstallationOctokit(Number(installationId));

    const apiHeaders = { 'X-GitHub-Api-Version': '2022-11-28' };

    // 1) Resolve base branch HEAD sha
    const baseRef = await octokit.request('GET /repos/{owner}/{repo}/git/ref/{ref}', {
      owner,
      repo,
      ref: `heads/${baseBranch}`,
      headers: apiHeaders
    });
    const baseSha = baseRef.data.object && baseRef.data.object.sha;
    if (!baseSha) {
      return res.status(500).send('Failed to resolve base branch HEAD SHA');
    }

    // 2) Create a unique PR branch off base
    let prBranch = buildBranchName(prBranchPrefix);
    prBranch = await ensureNewBranch(octokit, owner, repo, prBranch, baseSha, apiHeaders);

    // 3) Determine current file sha on the new branch (if file exists there)
    let currentSha = null;
    try {
      const meta = await octokit.request('GET /repos/{owner}/{repo}/contents/{path}', {
        owner, repo, path: filePath, ref: prBranch, headers: apiHeaders
      });
      currentSha = meta.data.sha;
    } catch (e) {
      // 404 means the file doesn't exist on that branch, that's fine
      if (!e || e.status !== 404) throw e;
    }

    // 4) Commit file content on PR branch
    const contentB64 = Buffer.from(csv, 'utf8').toString('base64');
    await octokit.request('PUT /repos/{owner}/{repo}/contents/{path}', {
      owner,
      repo,
      path: filePath,
      message: 'Update birthdays.csv via GitHub App backend (PR)',
      content: contentB64,
      sha: currentSha || undefined,
      branch: prBranch,
      headers: apiHeaders
    });

    // 5) Open PR from PR branch into base
    const prTitle = 'Update birthdays.csv';
    const prBody = 'Automated update of birthdays.csv via the Birth App backend.';
    const pr = await octokit.request('POST /repos/{owner}/{repo}/pulls', {
      owner,
      repo,
      title: prTitle,
      head: prBranch,
      base: baseBranch,
      body: prBody,
      headers: apiHeaders
    });

    return res.status(200).json({
      pr_url: pr.data && pr.data.html_url,
      pr_number: pr.data && pr.data.number,
      branch: prBranch
    });
  } catch (err) {
    console.error('GitHub App PR flow error:', err);
    return res.status(500).send('GitHub App PR flow error');
  }
};

function readRawBody(req) {
  return new Promise((resolve, reject) => {
    let body = '';
    req.on('data', (chunk) => (body += chunk));
    req.on('end', () => resolve(body));
    req.on('error', reject);
  });
}

function buildBranchName(prefix) {
  const stamp = new Date().toISOString().replace(/[-:TZ]/g, '').slice(0, 14);
  return `${prefix}-${stamp}`;
}

async function ensureNewBranch(octokit, owner, repo, name, baseSha, headers) {
  let branchName = name;
  let attempt = 0;
  while (true) {
    try {
      await octokit.request('POST /repos/{owner}/{repo}/git/refs', {
        owner,
        repo,
        ref: `refs/heads/${branchName}`,
        sha: baseSha,
        headers
      });
      return branchName;
    } catch (e) {
      // 422 Unprocessable Entity if ref already exists
      if (e && e.status === 422 && attempt < 5) {
        attempt += 1;
        const rand = Math.random().toString(36).slice(2, 8);
        branchName = `${name}-${rand}`;
        continue;
      }
      throw e;
    }
  }
}

File: api/_github.py
import base64
import json
import os
import time
import urllib.request
import urllib.error
import urllib.parse
from typing import Optional, Tuple

# Environment configuration with sensible defaults
GITHUB_OWNER = os.getenv("GITHUB_REPO_OWNER", "grinwi")
GITHUB_REPO = os.getenv("GITHUB_REPO", "birth-app")
GITHUB_BRANCH = os.getenv("GITHUB_BRANCH", "main")
GITHUB_FILE_PATH = os.getenv("GITHUB_FILE_PATH", "birthdays.csv")
GITHUB_TOKEN = os.getenv("GITHUB_TOKEN") or os.getenv("BIRTHDAY_APP_EDIT_CSV_TOKEN")

GITHUB_API_BASE = "https://api.github.com"
RAW_BASE = "https://raw.githubusercontent.com"


def _headers_json() -> dict:
    if not GITHUB_TOKEN:
        raise RuntimeError("Missing GITHUB_TOKEN or BIRTHDAY_APP_EDIT_CSV_TOKEN")
    return {
        "Authorization": f"Bearer {GITHUB_TOKEN}",
        "Accept": "application/vnd.github+json",
        "User-Agent": "birthdays-app-python",
        "Content-Type": "application/json",
        "X-GitHub-Api-Version": "2022-11-28",
    }


def github_request(url: str, method: str = "GET", data: Optional[dict] = None, headers: Optional[dict] = None):
    body = None
    if data is not None:
        body = json.dumps(data).encode("utf-8")
    req = urllib.request.Request(url, data=body, method=method)
    for k, v in (headers or {}).items():
        req.add_header(k, v)
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            return resp.getcode(), resp.read()
    except urllib.error.HTTPError as e:
        return e.code, e.read()
    except urllib.error.URLError as e:
        raise RuntimeError(f"Network error calling {url}: {e}")


def get_base_sha(owner: str, repo: str, branch: str) -> str:
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/git/ref/heads/{branch}"
    status, resp = github_request(url, headers=_headers_json())
    if status != 200:
        raise RuntimeError(f"Failed to resolve base ref: {status} {resp.decode('utf-8', 'ignore')}")
    data = json.loads(resp.decode("utf-8"))
    sha = (data.get("object") or {}).get("sha")
    if not sha:
        raise RuntimeError("Base branch SHA not found")
    return sha


def create_branch(owner: str, repo: str, base_sha: str, preferred_name: Optional[str] = None) -> str:
    # Try to create a unique branch name
    base = preferred_name or f"update-birthdays-{time.strftime('%Y%m%d%H%M%S')}"
    for i in range(6):
        name = base if i == 0 else f"{base}-{i}"
        url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/git/refs"
        status, resp = github_request(
            url,
            method="POST",
            data={"ref": f"refs/heads/{name}", "sha": base_sha},
            headers=_headers_json(),
        )
        if status == 201:
            return name
        # 422 means already exists, try next suffix
        if status != 422:
            raise RuntimeError(f"Failed to create branch: {status} {resp.decode('utf-8', 'ignore')}")
    raise RuntimeError("Could not create unique PR branch after multiple attempts")


def get_file_sha(owner: str, repo: str, path: str, ref: str) -> Optional[str]:
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/contents/{urllib.parse.quote(path)}?ref={urllib.parse.quote(ref)}"
    status, resp = github_request(url, headers=_headers_json())
    if status == 200:
        data = json.loads(resp.decode("utf-8"))
        return data.get("sha")
    if status == 404:
        return None
    raise RuntimeError(f"Failed to fetch file metadata: {status} {resp.decode('utf-8', 'ignore')}")


def put_file(owner: str, repo: str, path: str, branch: str, content_utf8: str, message: str, sha: Optional[str]) -> None:
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/contents/{urllib.parse.quote(path)}"
    payload = {
        "message": message,
        "content": base64.b64encode(content_utf8.encode("utf-8")).decode("ascii"),
        "branch": branch,
    }
    if sha:
        payload["sha"] = sha
    status, resp = github_request(url, method="PUT", data=payload, headers=_headers_json())
    if status not in (200, 201):
        raise RuntimeError(f"Failed to update file on GitHub: {status} {resp.decode('utf-8', 'ignore')}")


def open_pr(owner: str, repo: str, head: str, base: str, title: str, body: str) -> Tuple[int, str]:
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/pulls"
    status, resp = github_request(
        url,
        method="POST",
        data={"title": title, "head": head, "base": base, "body": body},
        headers=_headers_json(),
    )
    if status not in (200, 201):
        raise RuntimeError(f"Failed to open PR: {status} {resp.decode('utf-8', 'ignore')}")
    data = json.loads(resp.decode("utf-8"))
    return data.get("number"), data.get("html_url")


def fetch_raw_csv(owner: str, repo: str, branch: str, path: str) -> str:
    url = f"{RAW_BASE}/{owner}/{repo}/{branch}/{path}"
    req = urllib.request.Request(url, method="GET", headers={"User-Agent": "birthdays-app-python"})
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            return resp.read().decode("utf-8")
    except urllib.error.HTTPError as e:
        if e.code == 404:
            # Treat missing as empty CSV with header
            return ""
        raise
    except urllib.error.URLError as e:
        raise RuntimeError(f"Network error fetching raw CSV: {e}")


def create_pr_with_csv(new_csv_text: str, title: str, body: str = "") -> Tuple[int, str]:
    owner = GITHUB_OWNER
    repo = GITHUB_REPO
    base = GITHUB_BRANCH
    path = GITHUB_FILE_PATH

    base_sha = get_base_sha(owner, repo, base)
    branch_name = create_branch(owner, repo, base_sha, preferred_name=f"update-birthdays-{time.strftime('%Y%m%d%H%M%S')}")
    file_sha = get_file_sha(owner, repo, path, branch_name)
    put_file(owner, repo, path, branch_name, new_csv_text, message=title, sha=file_sha)
    pr_number, pr_url = open_pr(owner, repo, branch_name, base, title=title, body=body or "Automated update of birthdays.csv")
    return pr_number, pr_url
